.changeset/long-apes-admire/changes.md

```diff
@@ -0,0 +1,38 @@
+TypeScript types have been restructured. These changes:
+
+- Reduce build times when using emotion
+- In many cases remove the need for manually specifying generic parameters for your emotion components.
+
+If you encounter build issues after upgrade, try removing any manually specified generic types and let them be inferred. Otherwise refer to the breaking changes list below.
+
+## Improvements
+
+- useTheme added to EmotionTheming interface and can now create your own closed variation of withTheme. More information in the docs under the theming section.
+- Union types as props are better supported and should be inferred properly
+- Build times should be reduced significantly in larger projects.
+
+## Breaking changes
+
+- Function interpolation can no longer return another function directly
+- withTheme can now have the Theme type specified when calling it. For example `withTheme<MyTheme>(MyComponent)`
+
+  **Breaking change:** Generic argument changed, if you were specifying the ComponentType you will need to remove the generic parameter. Recommend following example setup in the TypeScript docs under theming section
+
+- `css` function has been restricted to prevent passing of invalid types
+- `CreateStyled` functions no longer take a second `ExtraProps` argument. Instead move it to after the create styled call. For example
+
+  `styled<typeof MyComponent, ExtraProps>(MyComponent)({})`  
+  to  
+  `styled(MyComponent)<ExtraProps>({})`
```

Comment:
is this 100% equivalent? could a codemod be written to ease the migration?

Reply:
In this scenario yes, there may be scenarios i've missed. 

Related to pain of upgrading, have you thought about moving the types to definitely typed. That way people could upgrade the types independent of the library.

Reply:
Do u know how the development on DefinitelyTyped works? Who can merge? Who makes the decisions. How to propose a change? And how to land such a big refactor as this? 

Migrating them there would certainly make our lives here easier, but I'm wondering what about quality for types there? Especially for a complex library like emotion?

Reply:
Yeah not sure, I think the TypeScript team merge. But they are working on automation around this using the bot.

It automatically will ping people who authored the types I think as well. Just a bit of an idea, not 100% how it would all work

Reply:
Regarding codemods, I hope not too many people are using the former syntax. Most of these breaking changes would be when people couldn't get the existing types to infer their props properly and had to be explicit. Or due to the type structures you needed to specify more generic args before.